frontend/src/app/main/data/workspace/variants.cljs

```diff
@@ -0,0 +1,229 @@
+    ptk/WatchEvent
+    (watch [it state _]
+      (let [page-id (:current-page-id state)
+            data    (dsh/lookup-file-data state)
+            objects    (dsh/lookup-page-objects state page-id)
+
+            related-components (->> (:components data)
+                                    vals
+                                    (filter #(= (:variant-id %) variant-id))
+                                    reverse)
+
+
+            property-name (str "Property" (-> related-components
+                                              first
+                                              :variant-properties
+                                              count
+                                              inc))
+
+            changes (-> (pcb/empty-changes it page-id)
+                        (pcb/with-library-data data)
+                        (pcb/with-objects objects))
+
+            [_ changes]
+            (reduce (fn [[num changes] component]
+                      (let [props        (-> (or (:variant-properties component) [])
+                                             (conj {:name property-name :value (str "Value" num)}))
+                            main-id      (:main-instance-id component)
+                            variant-name (properties-to-name props)]
+                        [(inc num)
+                         (-> changes
+                             (pcb/update-component (:id component) #(assoc % :variant-properties props))
+                             (pcb/update-shapes [main-id] #(assoc % :variant-name variant-name)))]))
+                    [1 changes]
+                    related-components)
+            undo-id (js/Symbol)]
+        (rx/of
+         (dwu/start-undo-transaction undo-id)
+         (dch/commit-changes changes)
+         (dwu/commit-undo-transaction undo-id))))))
+
+(defn set-variant-id
+  "Sets the variant-id on a component"
+  [component-id variant-id]
+  (ptk/reify ::set-variant-id
+    ptk/WatchEvent
+    (watch [it state _]
+      (let [page-id      (:current-page-id state)
+            data    (dsh/lookup-file-data state)
+            changes (-> (pcb/empty-changes it page-id)
+                        (pcb/with-library-data data)
+                        (pcb/update-component component-id #(assoc % :variant-id variant-id)))
+            undo-id  (js/Symbol)]
+        (rx/of
+         (dwu/start-undo-transaction undo-id)
+         (dch/commit-changes changes)
+         (dwu/commit-undo-transaction undo-id))))))
+
+(defn transform-in-variant
+  "Given the id of a main shape of a component, creates a variant structure for
+   that component"
+  [id]
+  (ptk/reify ::transform-in-variant
+    ptk/WatchEvent
+    (watch [_ state _]
+      (let [variant-id (uuid/next)
+            new-component-id (uuid/next)
+            file-id      (:current-file-id state)
+            page-id      (:current-page-id state)
+            objects      (dsh/lookup-page-objects state page-id)
```

Comment:
Pass `file-id` here as second argument, for avoid a duplicate lookup of the current-file-id